From dd3ab9bbf72b71bffd7e54e345f3793ac99b4a19 Mon Sep 17 00:00:00 2001
From: Ravi B <ravibabu@ti.com>
Date: Tue, 30 Aug 2011 18:26:37 +0530
Subject: [PATCH 0539/1092] musb: pm workaround fix in host mode when no
 device connected to ports

In host mode, the musb does not go to suspend state when no
device connected to port0/port1, only in the otg mode the try
idle function tries to turn off the session when no device is
connected to port.
This patch workaround this issue by clear/set the session bit
when system wide suspend/resume is invoked.

Signed-off-by: Ravi B <ravibabu@ti.com>
Signed-off-by: Vaibhav Hiremath <hvaibhav@ti.com>
---
 drivers/usb/musb/musb_host.c | 19 +++++++++++++++++++
 1 file changed, 19 insertions(+)

diff --git a/drivers/usb/musb/musb_host.c b/drivers/usb/musb/musb_host.c
index 382482c..66032cf 100644
--- a/drivers/usb/musb/musb_host.c
+++ b/drivers/usb/musb/musb_host.c
@@ -2364,6 +2364,14 @@ static int musb_bus_suspend(struct usb_hcd *hcd)
 	struct musb	*musb = hcd_to_musb(hcd);
 	u8		devctl;
 
+	if (!is_otg_enabled(musb) && is_host_enabled(musb)) {
+		devctl = musb_readb(musb->mregs, MUSB_DEVCTL);
+		devctl &= MUSB_DEVCTL_SESSION;
+		musb_writeb(musb->mregs, MUSB_DEVCTL, devctl);
+		musb->is_active = 0;
+		return 0;
+	}
+
 	if (!is_host_active(musb))
 		return 0;
 
@@ -2393,6 +2401,17 @@ static int musb_bus_suspend(struct usb_hcd *hcd)
 
 static int musb_bus_resume(struct usb_hcd *hcd)
 {
+	struct musb	*musb = hcd_to_musb(hcd);
+	u8		devctl;
+
+	if (!is_otg_enabled(musb) && is_host_enabled(musb)) {
+		devctl = musb_readb(musb->mregs, MUSB_DEVCTL);
+		devctl |= MUSB_DEVCTL_SESSION;
+		musb_writeb(musb->mregs, MUSB_DEVCTL, devctl);
+		musb->is_active = 1;
+		return 0;
+	}
+
 	/* resuming child port does the work */
 	return 0;
 }
-- 
1.7.11.2

